### ISSUE TITLE
Need a workaround for Yoga refusing to build with fastmath

### ISSUE THREAD

**Issue by julianstorer** [nick-thompson/blueprint#105]
Hit a bit of a gnarly problem, trying to add blueprint to Waveform. 

We build with fast-math enabled. Yoga refuses to work with that.

The messy part here is that it detects the __FASTMATH__ definition in the yoga _headers_, which are included by blueprint.h, and which therefore get pulled into the entire project. A decent workaround would be to disable the fast-math flag for just the blueprint cpp file, but that's not going to work because the headers will still fail in all the other TUs.

..which raises the question of why the duktape and yoga headers are included publicly in blueprint.h. Whenever we've embedded 3rd party libs like that in juce, we make a point of keeping them locked away to avoid potentially hundreds of user project files pulling them in when they're not actually needed.

Assuming that there's no reason why user-code would need to directly use any of the yoga/duktape types, it'd be a big improvement to both encapsulation and build times to get them out of the header, and use things like the Pimpl idiom where necessary to make the existing classes build.

This would also mean that we could either disable fast-math for just the blueprint.cpp file, or use some #pragmas inside that file, around the yoga includes, to selectively disable optimisation for that library

**Comment by julianstorer** [comment]
..and just another note on just what a pain this is..

Looking for simple workarounds, the code does seem to run OK if you comment-out the yoga check and let it compile with fast-math on. Maybe there are layouts where the maths will go wrong, but it's probably something we could live with.

However, to hack it, this  line need to be removed:

```
#if defined(__FAST_MATH__)
#error __FAST_MATH__ defined, refusing to compile
#endif
```

And because that's in the yoga submodule, the only way I could possibly commit this to our project would be to have our own fork of blueprint which points to our own hacked fork of yoga, which is just a mess.

Another hack would be to `#undef __FAST_MATH__` just before this code gets included, but then the same problem I mentioned above comes into play, because this header gets pulled into ALL translation units, so this undef would affect code that does need it to be correct.

**Comment by julianstorer** [comment]
Ah, hang on, I was getting yoga and duktape mixed up - the problematic check is in duk_config.h, and you seem to be including that directly, not via a submodule, so I can at least hack just my own fork of blueprint to bodge something that will compile. Whether it'll actually run all JS correctly is another matter of course..

All the above still stands though! Definitely think you should refactor to remove all these 3rd-party headers from the blueprint headers. There will probably be many projects like Waveform where compile times are already painful, and the fact that every translation unit will now drag in dozens more unused headers could be a deal-breaker for some people.

**Comment by nick-thompson** [comment]
For sure, thanks @julianstorer. This is really a design desicion borne by my ignorance in C++ dependency management. The goal in including these dependencies directly was to ensure that users of the blueprint juce module could just "add module" in their projucer or CMake projects and immediately be up and running. I realize this creates a configuration nightmare, which is partially captured in #62.

So, I agree this should be cleaned up, but I'll admit I'm not sure how best to do that. I'd appreciate some insight if you have it, and I'll go ahead and add the "help wanted" label here.

For clarity, I suppose Yoga would be easy enough: we could keep yoga as a submodule and put a `Pimpl` inside the `ShadowView` class implementation, which is the only place we're touching Yoga. Then using pragmas to fudge the preprocessor defs is plenty easy. For Duktape, I don't know really. It has more of a DIY integration approach, and I'm unsure what strategies we can use to make that build cleanly while still being easy for the end user.

**Comment by JoshMarler** [comment]
Happy to assist with this if you want to run over some ideas Nick. Was planning on expanding our cmake support so that blueprint could be added simply using the new cmake functions in juce without requiring user projects to add additional include paths.

This could very well all fall under a little "module restructure" ticket. 

I'd be happy to look into user configured duktape builds as that's some were gonna need at my shop. I've also got a hack of the blueprint duk_config.h in my fork at the moment. Be nice to streamline all this.

**Comment by nick-thompson** [comment]
Yes please! That'd be great

**Comment by nick-thompson** [comment]
After #145, I'm going to go ahead and mark this issue as a duplicate of #62. We should now be in a place where end users can provide their own Duktape source, modified however they need, and just specify some include defs to convince Blueprint to use that instead of the Duktape source that we bundle.

The work is done, but there are some lingering questions about how to most cleanly allow users to override the default-included Duktape source, which we'll address in #62
